parlai/core/agents.py

```diff
@@ -333,3 +233,109 @@ def report(self):
+    shared_agents = []
+    for shared_agent in shared:
+        agent = create_agent_from_shared(shared_agent)
+        shared_agents.append(agent)
+    return shared_agents
+
+def create_task_agent_from_taskname(opt):
+    """Creates task agent(s) assuming the input ``task_dir:teacher_class``.
+
+    e.g. def_string is a shorthand path like ``babi:Task1k:1`` or ``#babi``
+    or a complete path like ``parlai.tasks.babi.agents:Task1kTeacher:1``, 
+    which essentially performs ``from parlai.tasks.babi import Task1kTeacher``
+    with the parameter ``1`` in ``opt['task']`` to be used by the class ``Task1kTeacher``.
+    """
+    if ',' not in opt['task']:
+        # Single task
+        sp = opt['task'].strip().split(':')
+        if '.' in sp[0]:
+            module_name = sp[0]
+        else:
+            task = sp[0].lower()
+            module_name = "parlai.tasks.%s.agents" % (task)
+        if len(sp) > 1:
+            sp[1] = sp[1][0].upper() + sp[1][1:]
+            teacher = sp[1]
+            if '.' not in sp[0] and 'Teacher' not in teacher:
+                # Append "Teacher" to class name by default if
+                # a complete path is not given.
+                teacher += "Teacher"
+        else:
+            teacher = "DefaultTeacher"
+        my_module = importlib.import_module(module_name)
+        teacher_class = getattr(my_module, teacher)
+        task_agents = teacher_class(opt)
+        if type(task_agents) != list:
+            task_agents = [task_agents]
+        return task_agents
+    else:
+        # Multitask teacher/agent
+        task_agents = MultiTaskTeacher(opt)
+        if type(task_agents) != list:
+            task_agents = [task_agents]
+        return task_agents
+
+
+def _create_task_agents(opt):
+    """Creates task agent(s) for the given task name.
+    It does this by calling the create_agent function in agents.py of the
+    given task.
+    If create_agents function does not exist, it just looks for
+    the teacher (agent) class defined by the task name directly.
+    (This saves the task creator bothering to define the
+    create_agents function when it is not needed.)
+    """
+    sp = opt['task'].strip().split(':')
+    if '.' in sp[0]:
+        # The case of opt['task'] = 'parlai.tasks.squad.agents:DefaultTeacher'
+        # (i.e. specifying your own path directly)
+        module_name = sp[0]
+    else:
+        task = sp[0].lower()
+        module_name = "parlai.tasks.%s.agents" % (task)
+    my_module = importlib.import_module(module_name)
+    try:
+        # Tries to call the create_agent function in agents.py
+        create_agent = getattr(my_module, 'create_agents')
+        task_agents = create_agent(opt)
+    except AttributeError:
+        # Create_agent not found, so try to create the teacher directly.
+        return create_task_agent_from_taskname(opt)
+    if type(task_agents) != list:
+        task_agents = [task_agents]
+    return task_agents
```

Comment:
Yes it is - moving them around a little bit so that in the doc all of the classes will show first and then the methods :)